helper/communicator/step_connect_winrm.go

```diff
@@ -189,3 +196,22 @@ func (s *StepConnectWinRM) waitForWinRM(state multistep.StateBag, ctx context.Co
 
 	return comm, nil
 }
+
+// setNoProxy configures the $NO_PROXY env var
+func setNoProxy(host string, port int) error {
+	current := os.Getenv("NO_PROXY")
+	p := fmt.Sprintf("%s:%d", host, port)
```

Comment:
In testing this functionality out using an `amazon-ebs` builder and squid as my local proxy I found that Go does not honor IP addresses in the NO_PROXY environment variable; Even though it looks like it should https://github.com/golang/net/blob/3c3fba18258b2a1398a025a6aeb7374d2a470009/http/httpproxy/proxy.go#L237 :thinking: 

I found that it must be a domain name (FQDN or domain suffix with a glob *.example.com) with an optional port in order for the NO_PROXY value to kick in. Setting `ssh_interface` to `public_dns` worked as expected. 

Using Public DNS
```
2020/05/29 10:10:42 packer-builder-amazon-ebs plugin: Using host value: ec2-34-203-77-185.compute-1.amazonaws.com
2020/05/29 10:10:42 packer-builder-amazon-ebs plugin: Setting NO_PROXY environment variable
2020/05/29 10:10:42 packer-builder-amazon-ebs plugin: Setting NO_PROXY environment variable ec2-34-203-77-185.compute-1.amazonaws.com:5985
2020/05/29 10:10:42 packer-builder-amazon-ebs plugin: [INFO] Attempting WinRM connection...
2020/05/29 10:10:42 packer-builder-amazon-ebs plugin: [DEBUG] connecting to remote shell using WinRM
2020/05/29 10:10:49 packer-builder-amazon-ebs plugin: Checking that WinRM is connected with: 'powershell.exe -EncodedCommand aQBmACAAKABUAGUAcwB0AC0AUABhAHQAaAAgAHYAYQByAGkAYQBiAGwAZQA6AGcAbABvAGIAYQBsADoAUAByAG8AZwByAGUAcwBzAFAAcgBlAGYAZQByAGUAbgBjAGUAKQB7ACQAUAByAG8AZwByAGUAcwBzAFAAcgBlAGYAZQByAGUAbgBjAGUAPQAnAFMAaQBsAGUAbgB0AGwAeQBDAG8AbgB0AGkAbgB1AGUAJwB9ADsAIABlAGMAaABvACAAIgBXAGkAbgBSAE0AIABjAG8AbgBuAGUAYwB0AGUAZAAuACIA'
2020/05/29 10:10:49 packer-builder-amazon-ebs plugin: [INFO] starting remote command: powershell.exe -EncodedCommand aQBmACAAKABUAGUAcwB0AC0AUABhAHQAaAAgAHYAYQByAGkAYQBiAGwAZQA6AGcAbABvAGIAYQBsADoAUAByAG8AZwByAGUAcwBzAFAAcgBlAGYAZQByAGUAbgBjAGUAKQB7ACQAUAByAG8AZwByAGUAcwBzAFAAcgBlAGYAZQByAGUAbgBjAGUAPQAnAFMAaQBsAGUAbgB0AGwAeQBDAG8AbgB0AGkAbgB1AGUAJwB9ADsAIABlAGMAaABvACAAIgBXAGkAbgBSAE0AIABjAG8AbgBuAGUAYwB0AGUAZAAuACIA
2020/05/29 10:10:49 packer-builder-amazon-ebs plugin: [INFO] command 'powershell.exe -EncodedCommand aQBmACAAKABUAGUAcwB0AC0AUABhAHQAaAAgAHYAYQByAGkAYQBiAGwAZQA6AGcAbABvAGIAYQBsADoAUAByAG8AZwByAGUAcwBzAFAAcgBlAGYAZQByAGUAbgBjAGUAKQB7ACQAUAByAG8AZwByAGUAcwBzAFAAcgBlAGYAZQByAGUAbgBjAGUAPQAnAFMAaQBsAGUAbgB0AGwAeQBDAG8AbgB0AGkAbgB1AGUAJwB9ADsAIABlAGMAaABvACAAIgBXAGkAbgBSAE0AIABjAG8AbgBuAGUAYwB0AGUAZAAuACIA' exited with code: 0
    amazon-ebs: WinRM connected.
2020/05/29 10:10:49 packer-builder-amazon-ebs plugin: Connected to machine
==> amazon-ebs: Connected to WinRM!
2020/05/29 10:10:49 packer-builder-amazon-ebs plugin: Running the provision hook
```

Using Public IP
```
2020/05/29 10:15:31 packer-builder-amazon-ebs plugin: Using host value: 52.87.222.97                                                               
2020/05/29 10:15:31 packer-builder-amazon-ebs plugin: Setting NO_PROXY environment variable                                                        
2020/05/29 10:15:31 packer-builder-amazon-ebs plugin: Setting NO_PROXY environment variable 52.87.222.97:5985                                      
2020/05/29 10:15:31 packer-builder-amazon-ebs plugin: [INFO] Attempting WinRM connection...                                                        
2020/05/29 10:15:31 packer-builder-amazon-ebs plugin: [DEBUG] connecting to remote shell using WinRM                                               
2020/05/29 10:15:31 packer-builder-amazon-ebs plugin: [ERROR] connection error: unknown error Post "http://52.87.222.97:5985/wsman": proxyconnect t
cp: dial tcp 172.16.103.130:80: connect: connection refused                                                                                        
2020/05/29 10:15:31 packer-builder-amazon-ebs plugin: [ERROR] WinRM connection err: unknown error Post "http://52.87.222.97:5985/wsman": proxyconne
ct tcp: dial tcp 172.16.103.130:80: connect: connection refused 
```

Reply:
https://github.com/golang/go/issues/2158

Reply:
The above issue suggests that they know it doesn't work for IP addresses, and don't actually care. But I see in someone's sharing of a python implementation something that may be useful: 

https://github.com/golang/go/issues/2158#issuecomment-66058156

If instead of using the IP of the instance we just set no_proxy = "*", would that work? Since the winrm commands are being executed in their own shell environment, it shouldn't be a problem to just brute force the proxy to turn off for all winrm connections.

Reply:
looks to me like https://github.com/golang/net/blob/3c3fba18258b2a1398a025a6aeb7374d2a470009/http/httpproxy/proxy.go#L224 implements something similar, so I feel like it could work.

Reply:
I think the "*" could work. I didn't get a chance to test out that option as I was playing around with different IP options. But happy to test since I still have a squid env running.

Reply:
That would be super helpful. I haven't spun up a repro env yet.

Reply:
Okay so I suspect something else is going on. I tested with "*" and its a no go, I also found that my local NO_PROXY `NO_PROXY=*.amazonaws.com` were not getting stomped out by the builder setting NO_PROXY to just the dns of the compute instance, which is why it worked when using public DNS.

After running with the following configuration all test failed so maybe we should get a second set of eyes here

```
export NO_PROXY=*.us-east-1.amazonaws.com

Update build.json (any amazon ebs windows build)
set "winrm_no_proxy": true
packer build build.json 
```

According to these tests it should work https://github.com/golang/net/blob/3c3fba18258b2a1398a025a6aeb7374d2a470009/http/httpproxy/proxy_test.go#L261

I did a quick check to see if the winrm pkg was getting the updated NO_PROXY setting and it is so not sure what is going on. Maybe I have a bad setup.

Reply:
Okay so I did a little more testing to try an understand why setting NO_PROXY to "*" was not working and found this issue https://github.com/golang/go/issues/22079 implying that Proxy settings are only read once. Which lead me to think maybe our winrm client is being established before the NO_PROXY environment variable is updated so our changes never get picked up. 

That line of thinking lead me to [StepConnect](https://github.com/hashicorp/packer/blob/master/helper/communicator/step_connect.go#L71) then finally to [PrepareWinRM](https://github.com/hashicorp/packer/blob/793b425dce1bd70e48acfe0374d17555019077a2/helper/communicator/config.go#L499). I dropped the following code snippet in the PrepareWinRM method on line [509](https://github.com/hashicorp/packer/blob/793b425dce1bd70e48acfe0374d17555019077a2/helper/communicator/config.go#L509) and found that I am now able to connect when setting `"winrm_no_proxy": true`.
 
```go
       if c.WinRMNoProxy {
               err := os.Setenv("NO_PROXY", "*")
               if err != nil {
                       log.Printf("[DEBUG] failed in setting NO_PROXY %s", err)
               }
       }
```

<details>
<summary>My testing environment</summary>

Squid Proxy running in a vagrant box using an standard install with an extra deny statement for ports 5985 and 5986. 


Proxy variables set on my localhost running Packer
```
HTTP_PROXY=172.16.103.130
HTTPS_PROXY=172.16.103.130
NO_PROXY=*.us-east-1.amazonaws.com //needed in my setup to allow AWS api calls
```

Packer from https://www.packer.io/intro/getting-started/build-image/#a-windows-example with a couple of modifications

```
{
  "variables": {
    "aws_access_key": "{{env `AWS_ACCESS_KEY_ID`}}",
    "aws_secret_key": "{{env `AWS_SECRET_ACCESS_KEY`}}",
    "region":         "us-east-1"
  },
  "builders": [
    {
      "type": "amazon-ebs",
      "access_key": "{{ user `aws_access_key` }}",
      "secret_key": "{{ user `aws_secret_key` }}",
      "region": "{{ user `region` }}",
      "instance_type": "t2.micro",
      "source_ami_filter": {
        "filters": {
          "virtualization-type": "hvm",
          "name": "*Windows_Server-2012-R2*English-64Bit-Base*",
          "root-device-type": "ebs"
        },
        "most_recent": true,
        "owners": "amazon"
      },
      "ami_name": "packer-demo-{{timestamp}}",
      "user_data_file": "./bootstrap_win.txt",
      "communicator": "winrm",
      "winrm_no_proxy": true,
      "winrm_timeout": "3m",
      "winrm_username": "Administrator",
      "winrm_password": "SuperS3cr3t!!!!"
    }
  ],
  "provisioners": [
    {
      "type": "powershell",
       "inline": ["dir c:/Windows/Temp/"]
    },
  ]
}
```
</details>

<details>
<summary> Results</summary>

`winrm_no_proxy` set to true
```
==> amazon-ebs: Waiting for WinRM to become available...                                                                                                                                                                                                                                               
2020/05/30 05:58:13 packer-builder-amazon-ebs plugin: Using host value: 52.91.136.238                                                                                                                                                                                                                  
2020/05/30 05:58:13 packer-builder-amazon-ebs plugin: [INFO] Attempting WinRM connection...                                                                                                                                                                                                            
2020/05/30 05:58:13 packer-builder-amazon-ebs plugin: [DEBUG] connecting to remote shell using WinRM                                                                                                                                                                                                   
2020/05/30 05:58:43 packer-builder-amazon-ebs plugin: [ERROR] connection error: unknown error Post "http://52.91.136.238:5985/wsman": dial tcp 52.91.136.238:5985: i/o timeout                                                                                                                         
2020/05/30 05:58:43 packer-builder-amazon-ebs plugin: [ERROR] WinRM connection err: unknown error Post "http://52.91.136.238:5985/wsman": dial tcp 52.91.136.238:5985: i/o timeout                                                                                                                     
2020/05/30 05:58:48 packer-builder-amazon-ebs plugin: Using host value: 52.91.136.238                                                                                                                                                                                                                  
2020/05/30 05:58:48 packer-builder-amazon-ebs plugin: [INFO] Attempting WinRM connection...                                                                                                                                                                                                            
2020/05/30 05:58:48 packer-builder-amazon-ebs plugin: [DEBUG] connecting to remote shell using WinRM                                                                                                                                                                                                   
2020/05/30 05:59:09 packer-builder-amazon-ebs plugin: Checking that WinRM is connected with: 'powershell.exe -EncodedCommand aQBmACAAKABUAGUAcwB0AC0AUABhAHQAaAAgAHYAYQByAGkAYQBiAGwAZQA6AGcAbABvAGIAYQBsADoAUAByAG8AZwByAGUAcwBzAFAAcgBlAGYAZQByAGUAbgBjAGUAKQB7ACQAUAByAG8AZwByAGUAcwBzAFAAcgBlAGYAZQ
ByAGUAbgBjAGUAPQAnAFMAaQBsAGUAbgB0AGwAeQBDAG8AbgB0AGkAbgB1AGUAJwB9ADsAIABlAGMAaABvACAAIgBXAGkAbgBSAE0AIABjAG8AbgBuAGUAYwB0AGUAZAAuACIA'                                                                                                                                                                
2020/05/30 05:59:09 packer-builder-amazon-ebs plugin: [INFO] starting remote command: powershell.exe -EncodedCommand aQBmACAAKABUAGUAcwB0AC0AUABhAHQAaAAgAHYAYQByAGkAYQBiAGwAZQA6AGcAbABvAGIAYQBsADoAUAByAG8AZwByAGUAcwBzAFAAcgBlAGYAZQByAGUAbgBjAGUAKQB7ACQAUAByAG8AZwByAGUAcwBzAFAAcgBlAGYAZQByAGUAbg
BjAGUAPQAnAFMAaQBsAGUAbgB0AGwAeQBDAG8AbgB0AGkAbgB1AGUAJwB9ADsAIABlAGMAaABvACAAIgBXAGkAbgBSAE0AIABjAG8AbgBuAGUAYwB0AGUAZAAuACIA                                                                                                                                                                         
2020/05/30 05:59:11 packer-builder-amazon-ebs plugin: [INFO] command 'powershell.exe -EncodedCommand aQBmACAAKABUAGUAcwB0AC0AUABhAHQAaAAgAHYAYQByAGkAYQBiAGwAZQA6AGcAbABvAGIAYQBsADoAUAByAG8AZwByAGUAcwBzAFAAcgBlAGYAZQByAGUAbgBjAGUAKQB7ACQAUAByAG8AZwByAGUAcwBzAFAAcgBlAGYAZQByAGUAbgBjAGUAPQAnAFMAaQ
BsAGUAbgB0AGwAeQBDAG8AbgB0AGkAbgB1AGUAJwB9ADsAIABlAGMAaABvACAAIgBXAGkAbgBSAE0AIABjAG8AbgBuAGUAYwB0AGUAZAAuACIA' exited with code: 0                                                                                                                                                                    
    amazon-ebs: WinRM connected.
```

`winrm_no_proxy` set to false
```
2020/05/30 06:38:01 packer-builder-amazon-ebs plugin: Waiting for WinRM, up to timeout: 3m0s
2020/05/30 06:38:01 packer-builder-amazon-ebs plugin: Using host value: 18.210.10.78
2020/05/30 06:38:01 packer-builder-amazon-ebs plugin: [INFO] Attempting WinRM connection...
2020/05/30 06:38:01 packer-builder-amazon-ebs plugin: [DEBUG] connecting to remote shell using WinRM
2020/05/30 06:38:01 packer-builder-amazon-ebs plugin: [ERROR] connection error: unknown error Post "http://18.210.10.78:5985/wsman": proxyconnect tcp: dial tcp 172.16.103.130:80: connect: connection refused
2020/05/30 06:38:01 packer-builder-amazon-ebs plugin: [ERROR] WinRM connection err: unknown error Post "http://18.210.10.78:5985/wsman": proxyconnect tcp: dial tcp 172.16.103.130:80: connect: connection refused
2020/05/30 06:38:06 packer-builder-amazon-ebs plugin: Using host value: 18.210.10.78
2020/05/30 06:38:06 packer-builder-amazon-ebs plugin: [INFO] Attempting WinRM connection...
2020/05/30 06:38:06 packer-builder-amazon-ebs plugin: [DEBUG] connecting to remote shell using WinRM
2020/05/30 06:38:06 packer-builder-amazon-ebs plugin: [ERROR] connection error: unknown error Post "http://18.210.10.78:5985/wsman": proxyconnect tcp: dial tcp 172.16.103.130:80: connect: connection refused
2020/05/30 06:38:06 packer-builder-amazon-ebs plugin: [ERROR] WinRM connection err: unknown error Post "http://18.210.10.78:5985/wsman": proxyconnect tcp: dial tcp 172.16.103.130:80: connect: connection refused
2020/05/30 06:38:11 packer-builder-amazon-ebs plugin: Using host value: 18.210.10.78
2020/05/30 06:38:11 packer-builder-amazon-ebs plugin: [INFO] Attempting WinRM connection...
2020/05/30 06:38:11 packer-builder-amazon-ebs plugin: [DEBUG] connecting to remote shell using WinRM
2020/05/30 06:38:11 packer-builder-amazon-ebs plugin: [ERROR] connection error: unknown error Post "http://18.210.10.78:5985/wsman": proxyconnect tcp: dial tcp 172.16.103.130:80: connect: connection refused
2020/05/30 06:38:11 packer-builder-amazon-ebs plugin: [ERROR] WinRM connection err: unknown error Post "http://18.210.10.78:5985/wsman": proxyconnect tcp: dial tcp 172.16.103.130:80: connect: connection refused
2020/05/30 06:38:16 packer-builder-amazon-ebs plugin: Using host value: 18.210.10.78
```
</details>

 I'm not fully confident that I understand winrm code, or even squid :sweat_smile:, enough to say this is all that is needed, but seeing as the last round of tests are  passing I think it can help to set the NO_PROXY changes as early as possible.

Reply:
At the very least this seems like a very reasonable guess. Can you push your changes to this branch?

Reply:
I pushed up my changes. I used hard coded values just to show what I was testing with.

Reply:
This feels like it would turn off the proxy even for the API commands, no?

Reply:
Looking at that thread you linked it makes me think that the correct solution is to update the winrm vendored code to use https://godoc.org/golang.org/x/net/http/httpproxy instead of the ProxyFromEnvironment code.

Reply:
> Looking at that thread you linked it makes me think that the correct solution is to update the winrm vendored code to use https://godoc.org/golang.org/x/net/http/httpproxy instead of the ProxyFromEnvironment code.

Yes, but I saw two PRs from Hooker that did just that so that made me think that we were setting it too late https://github.com/masterzen/winrm/pull/66, https://github.com/masterzen/winrm/pull/70

Reply:
Sorry I hit enter too quick. I deleted my last comment. I think an upstream change could be the best way forward for sure.